wee_slack.py

```diff
@@ -1422,16 +1423,49 @@ def update_member_presence(self, user, presence):
                 c.update_nicklist(user.id)
 
     def subscribe_users_presence(self):
-        # FIXME: There is a limitation in the API to the size of the
-        # json we can send.
-        # We should try to be smarter to fetch the users whom we want to
-        # subscribe to.
-        users = list(self.users.keys())[0:750]
+        if len(self.users) <= PRESENCE_SUBSCRIBE_SIZE:
+            users = list(self.users.keys())
+        else:
+            users = self.get_users_by_presence_order()
+
         self.send_to_websocket({
             "type": "presence_sub",
             "ids": users,
         }, expect_reply=False)
 
+    def get_users_by_presence_order(self):
+        """
+        Get users in this order (restricted by the API JSON size limit):
+
+        1. Current user
+        2. Users in DMs
+        3. Users in MPDMs
+        4. Users in private groups
+        5. Users in public channels (excluding the "general" channel)
+        6. Users in the "general" channel
+        """
+        dm_users, mpdm_users, group_users, channel_users = set(), set(), set(), set()
+        for channel in self.channels.values():
+            if getattr(channel, "is_open", False):
+                if channel.type == "im":
+                    dm_users.add(channel.user)
+                elif channel.type == "mpim":
+                    mpdm_users.update(channel.members)
+                elif channel.type == "group":
+                    group_users.update(channel.members)
+            elif getattr(channel, "is_member", False):
+                if channel.type == "channel" and not channel.is_general:
+                    channel_users.update(channel.members)
```

Comment:
This can be shortened to:

```
        users = defaultdict(set)
        for channel in self.channels.values():
            if (getattr(channel, "is_open", False) or getattr(channel, "is_member", False)
                 and not getattr(channel, "is_general", False)):
                users[channel.type].update(channel.get_members())
```

Then use `users['im']` instead of `dm_users` etc. `defaultdict` comes from `collections`.

Reply:
Nice, though I think it'd be fine to just define a normal dict with channel types because they're known, e.g. `users = dict.fromkeys(("im", "mpim", "group", "channel", "shared"), set())`